database/assignment-1/script.sql

```diff
@@ -1,154 +1,178 @@
--- Active: 1677384415382@@127.0.0.1@8888@blog_app@public
+-- Active: 1677603582207@@127.0.0.1@8888@blog_app@public
 CREATE TYPE user_role AS ENUM('author', 'admin', 'moderator');
 
 CREATE TABLE users(
-    id serial NOT NULL PRIMARY KEY,
-    username VARCHAR(30) NOT NULL UNIQUE,
-    password VARCHAR(20) NOT NULL,
+    id SERIAL PRIMARY KEY,
+    email VARCHAR(30) NOT NULL UNIQUE,
+    password VARCHAR(50) NOT NULL,
     role user_role NOT NULL
 );
 
-CREATE TYPE post_status AS ENUM('draft', 'post');
+CREATE TABLE categories(
+    id SERIAL PRIMARY KEY,
+    name VARCHAR(50) NOT NULL UNIQUE,
+    description VARCHAR(100) NOT NULL
+);
+
+CREATE TYPE post_status AS ENUM('draft', 'published');
 
 CREATE TABLE posts(
-    id serial NOT NULL PRIMARY KEY,
+    id SERIAL PRIMARY KEY,
     user_id INT NOT NULL,
     Foreign Key (user_id) REFERENCES users(id) ON DELETE NO ACTION,
-    title VARCHAR(100) NOT NULL,
+    title VARCHAR(200) NOT NULL,
     content TEXT NOT NULL,
     status post_status NOT NULL,
-    created_at TIMESTAMP DEFAULT NOW()
+    category_id INT NOT NULL,
+    Foreign Key (category_id) REFERENCES categories(id) ON DELETE NO ACTION,
+    created_at TIMESTAMP DEFAULT NOW(),
+    deleted_at TIMESTAMP DEFAULT NOW()
 );
 
-CREATE TABLE categories(
-    post_id int NOT NULL,
-    Foreign Key (post_id) REFERENCES posts(id) ON DELETE NO ACTION,
-    name VARCHAR(30) NOT NULL UNIQUE,
+CREATE TABLE tags(
+    id SERIAL PRIMARY KEY,
+    name VARCHAR(50) NOT NULL UNIQUE,
     description VARCHAR(100) NOT NULL
 );
 
-CREATE TABLE tags(
-    post_id int NOT NULL,
+-- one post can have many tags and one tag can be associated with many posts
+CREATE TABLE post_tags(
+    post_id INT NOT NULL,
     Foreign Key (post_id) REFERENCES posts(id) ON DELETE NO ACTION,
-    name VARCHAR(30) NOT NULL UNIQUE,
-    description VARCHAR (100) NOT NULL
+    tag_id INT NOT NULL,
+    Foreign Key (tag_id) REFERENCES tags(id) ON DELETE NO ACTION,
+    PRIMARY KEY (post_id, tag_id)
 );
 
 CREATE TABLE comments(
-    id SERIAL NOT NULL PRIMARY KEY,
+    id SERIAL PRIMARY KEY,
     user_id INT NOT NULL,
-    post_id INT NOT NULL,
     Foreign Key (user_id) REFERENCES users(id) ON DELETE NO ACTION,
+    post_id INT NOT NULL,
     Foreign Key (post_id) REFERENCES posts(id) ON DELETE NO ACTION,
     comment TEXT NOT NULL,
+    parent_comment_id INT REFERENCES comments(id) ON DELETE NO ACTION,
     created_at TIMESTAMP DEFAULT NOW()
 );
 
-CREATE TABLE replies(
-    comment_id INT NOT NULL,
-    Foreign Key (comment_id) REFERENCES comments(id) ON DELETE NO ACTION,
-    reply TEXT NOT NULL,
-    created_at TIMESTAMP DEFAULT NOW()
-);
-
-CREATE TABLE post_metadata(
-    post_id INT NOT NULL,
+CREATE TABLE metadata(
+    id SERIAL PRIMARY KEY,
```

Comment:
still not scalable , what if we want to add new property of a post how do you handle it?